sound/soc/sof/ipc4-topology.c

```diff
@@ -1648,6 +1661,87 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int
+sof_ipc4_process_set_pin_formats(struct snd_sof_widget *swidget,
+				 struct sof_ipc4_pin_format *pin_format, int num_pins,
+				 int pin_type)
+{
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_soc_component *scomp = swidget->scomp;
+	int i, j;
+
+	for (i = 0; i < num_pins; i++, pin_format++) {
+		/* Pin 0 audio formats are derived from the base config input/output format */
+		if (i == 0) {
+			if (pin_type == SOF_PIN_TYPE_SINK) {
+				pin_format->buffer_size = process->base_config.ibs;
+				pin_format->audio_fmt = process->base_config.audio_fmt;
+			} else {
+				pin_format->buffer_size = process->base_config.obs;
+				pin_format->audio_fmt = process->output_format;
+			}
+			continue;
+		}
+
+		/* For all other pins, find the pin formats from those set in topology */
+		for (j = 0; j < 2 * SOF_WIDGET_MAX_NUM_PINS; j++) {
+			struct sof_ipc4_pin_info pin_info = process->pin_info[j];
+
+			if (pin_info.pin_type == pin_type && pin_info.fmt.pin_index == i) {
+				*pin_format = pin_info.fmt;
+				break;
+			}
+		}
+
+		if (j == 2 * SOF_WIDGET_MAX_NUM_PINS) {
+			dev_err(scomp->dev, "%s pin %d format not found for %s\n",
+				(pin_type == SOF_PIN_TYPE_SINK) ? "sink" : "source",
+				i, swidget->widget->name);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+static struct sof_ipc4_base_module_cfg_ext *
+sof_ipc4_process_add_base_cfg_extn(struct snd_sof_widget *swidget, size_t *size)
+{
+	struct sof_ipc4_base_module_cfg_ext *base_cfg_ext;
+	struct sof_ipc4_pin_format *pin_format;
+	int num_total_pins = swidget->num_sink_pins + swidget->num_source_pins;
+	int ret;
+
+	*size = struct_size(base_cfg_ext, pin_formats, num_total_pins);
+
+	base_cfg_ext = kzalloc(*size, GFP_KERNEL);
```

Comment:
The base_cfg_ext can be used to configure the pins of the module, yes. It can configure individual pins selectively.

If the module for example have 4 input and 2 output pins and only input pin 3 and output pin 0 is used, then the base_cfg_ext shall be:
```
u16 num_sink_pin_fmts = 1;
u16 num_source_pin_fmts = 1;
u8 reserved[12];
struct sof_ipc4_pin_format sink_pin_fmt[1];
struct sof_ipc4_pin_format source_pin_fmt[1];
```
Despite that the module have altogether 6 pins, the cfg_ext only have format for the two used pins.

By the current implementation, the sink_pin_fmt will contain the format from the base config (common for pin0s) and the format for the source pin would be coming as a individual section from tplg?

Reply:
@ujfalusi this implementation assumes that the token num_source_pins/num_sink_pins coming from topology is the number of pins that are actually ued by the processing module. so if the module supports a max of 4 and 3 input/output pins but only 1 input/output pin is used, the tplg must set the num_source_pins/num_sink_pins to 1/1.